### PR TITLE
Add cqm

### LINKED ISSUE #13: Add CQM demo file
Assign to @vgoliber please :)

### LINKED ISSUE #14: Canada case failing
Error message:

Reading shp file...
Traceback (most recent call last):
  File "/workspace/graph-coloring/map_coloring.py", line 157, in <module>
    state_records, state_neighbors = get_state_info(input_shp_file)
  File "/workspace/graph-coloring/map_coloring.py", line 49, in get_state_info
    for state in sf.records():
  File "/workspace/.pip-modules/lib/python3.7/site-packages/shapefile.py", line 1306, in records
    r = self.__record(oid=i)
  File "/workspace/.pip-modules/lib/python3.7/site-packages/shapefile.py", line 1281, in __record
    value = u(value, self.encoding, self.encodingErrors)
  File "/workspace/.pip-modules/lib/python3.7/site-packages/shapefile.py", line 104, in u
    return v.decode(encoding, encodingErrors)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe9 in position 11: invalid continuation byte

### PR DESCRIPTION
**Pull request by vgoliber**
Resolves #14 and resolves #13

### CODE REVIEW

File: graph_coloring.py [hunk 1]
```diff
@@ -12,78 +12,107 @@
...
+    # Add constraint to make variables discrete
+    for n in G.nodes():
+        cqm.add_discrete([(n,i) for i in range(num_colors)])
+  
+    # Build the constraints: edges have different color end points
+    for u, v in G.edges:
+        for i in range(num_colors):
+            c = BinaryQuadraticModel('BINARY')
+            c.set_quadratic((u,i),(v,i),1)
+            cqm.add_constraint(c == 0)
```
Comment (arcondello) [inline graph_coloring.py:63]: I am wondering for this example, if it makes sense to use the less performant, but more pedagogical symbolic construction
```python
colors = {v: {c: dimod.Binary((v, c)) for c in range(num_colors)} for v in graph.nodes}

for u, v in G.edges:
    for c in range(num_colors):
        cqm.add_constraint(colors[u][c]*colors[v][c] == 0)
```

Comment (vgoliber) [reply graph_coloring.py:63]: Are you thinking just on the graph coloring example and leave the map coloring one more performant for demos?

File: graph_coloring.py [hunk 2]
```diff
@@ -12,78 +12,104 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-import networkx as nx
-from dimod import DiscreteQuadraticModel
-from dwave.system import LeapHybridDQMSampler
 import matplotlib
+import networkx as nx
+from dimod import CQM, BinaryQuadraticModel, Binary, quicksum
```
Comment (mcfarljm) [inline graph_coloring.py:17]: Suggest using the full name:
```suggestion
from dimod import ConstrainedQuadraticModel, BinaryQuadraticModel, Binary, quicksum
```

File: map_coloring.py [hunk 3]
```diff
@@ -20,8 +20,8 @@
 import shapefile
 import matplotlib
 import networkx as nx
-from dimod import DiscreteQuadraticModel
-from dwave.system import LeapHybridDQMSampler
+from dimod import CQM, BinaryQuadraticModel
```
Comment (mcfarljm) [inline map_coloring.py:23]: Suggest using full name:
```suggestion
from dimod import ConstrainedQuadraticModel, BinaryQuadraticModel
```

File: graph_coloring.py [hunk 4]
```diff
@@ -12,78 +12,104 @@
...
+
+        pos (dict):
+            Plotting information for graph so that same graph shape is used.
+    """
+
+    print("\nProcessing sample...")
+
+    node_colors = [sample[i] for i in G.nodes()]
+    nx.draw(G, pos=pos, node_color=node_colors, node_size=50, edgecolors='k', cmap='hsv')
+    plt.savefig('graph_result.png')
```
Comment (mcfarljm) [inline graph_coloring.py:98]: Suggest adding a print statement like in `map_coloring.py`, so the user knows that a plot was created.

**Review by mcfarljm — APPROVED** [review]
LGTM, just a few minor suggestions.

File: README.md [hunk 5]
```diff
@@ -1,127 +1,97 @@
...
+
+### Constraints
+
+In this example, we have two groups of constraints.
+
+#### Discrete Constraints
+
+We defined our CQM using binary variables `(n, i)` to indicate assigning color
+`i` to node `n`. Since each node must be assigned exactly one color, we need to
+add a set of discrete constraints: one for each node `n`.
```
Comment (JoelPasvolsky) [inline README.md:71]: "one for each" --> ambiguous whether "one" refers to constraints or colors

File: README.md [hunk 6]
```diff
@@ -1,127 +1,97 @@
...
+
+#### Edge Constraints
+
+Our assignment of colors to nodes must satisfy the constraint that no edge has
+endpoints with the same color. In terms of map coloring, this is equivalent to
+no adjacent region having the same color assigned.
```
Comment (JoelPasvolsky) [inline README.md:77]: "In terms of map coloring, this is equivalent to no adjacent region" --> easier to parse something like "In terms of map coloring, this is equivalent to the constraint that no adjacent regions having the same colors assigned"

File: README.md [hunk 7]
```diff
@@ -1,127 +1,97 @@
...
+
+To build this constraint, we consider each edge `(u,v)` independently, and each
```
Comment (JoelPasvolsky) [inline README.md:79]: This paragraph is missing spaces after the commas of `(u, v)`

File: graph_coloring.py [hunk 8]
```diff
@@ -12,78 +12,107 @@
...
+def build_cqm(G, num_colors):
+    """Build CQM model."""
+
+    print("\nBuilding constrained quadratic model...")
+
+    # Initialize the CQM object
+    cqm = ConstrainedQuadraticModel()
+
+    # Build CQM variables
+    colors = {n: {c: Binary((n,c)) for c in range(num_colors)} for n in G.nodes}
```
Comment (JoelPasvolsky) [inline graph_coloring.py:48]: `Binary((n,c))` --> `Binary((n, c))`

File: graph_coloring.py [hunk 9]
```diff
@@ -12,78 +12,107 @@
...
+
+    # Build the objective: sum all colors values
+    cqm.set_objective(quicksum([c*colors[n][c] for n in G.nodes for c in range(num_colors)]))
+
+    # Add constraint to make variables discrete
+    for n in G.nodes():
+        cqm.add_discrete([(n,c) for c in range(num_colors)])
```
Comment (JoelPasvolsky) [inline graph_coloring.py:55]: `[(n,c)` --> `[(n, c)`

File: graph_coloring.py [hunk 10]
```diff
@@ -12,78 +12,107 @@
...
+    print("\nRunning hybrid sampler...")
+
+    # Initialize the CQM solver
+    sampler = LeapHybridCQMSampler()
+
+    # Solve the problem using the CQM solver
+    sampleset = sampler.sample_cqm(cqm, label='Example - Graph Coloring')
+    feasible_sampleset = sampleset.filter(lambda row: row.is_feasible)
+
+    ss = feasible_sampleset.first.sample
```
Comment (JoelPasvolsky) [inline graph_coloring.py:76]: What about nicer error handling for the (unlikely) case the solver fails to find a feasible solution rather than leaving it to break and return a Python error message?

Also, "ss" is not the best name in public example

File: graph_coloring.py [hunk 11]
```diff
@@ -12,78 +12,107 @@
...
+def build_graph(num_nodes):
+    """Build graph."""
+
+    print("\nBuilding graph...")
+
+    G = nx.powerlaw_cluster_graph(num_nodes, 3, 0.4)
+    pos = nx.spring_layout(G)
+    nx.draw(G, pos=pos, node_size=50, edgecolors='k', cmap='hsv')
+    plt.savefig("original_graph.png")
+    plt.clf()
```
Comment (JoelPasvolsky) [inline graph_coloring.py:35]: Why is this needed?

File: map_coloring.py [hunk 12]
```diff
@@ -69,45 +69,51 @@ def build_graph(state_neighbors):
...
+    # Initialize the CQM object
+    cqm = ConstrainedQuadraticModel()
 
-    # Initialize the DQM object
-    dqm = DiscreteQuadraticModel()
+    # Build the objective: sum all colors values
+    obj = BinaryQuadraticModel('BINARY')
+    for n in G.nodes():
+        for i in range(num_colors):
+            obj.set_linear((n,i), i)
```
Comment (JoelPasvolsky) [inline map_coloring.py:84]: `obj.set_linear((n,i), i)` --> `obj.set_linear((n, i), i)`

File: map_coloring.py [hunk 13]
```diff
@@ -69,45 +69,51 @@ def build_graph(state_neighbors):
...
-    # have the effect of minimizing the number of colors used.
-    # We penalize edge connections by the Lagrange parameter, to encourage
-    # connected nodes to have different colors.
-    for v in G.nodes:
-        dqm.add_variable(num_colors, label=v)
-    for v in G.nodes:
-        dqm.set_linear(v, colors)
+    # Add constraint to make variables discrete
+    for n in G.nodes():
+        cqm.add_discrete([(n,i) for i in range(num_colors)])
```
Comment (JoelPasvolsky) [inline map_coloring.py:89]: `cqm.add_discrete([(n,i)` --> `cqm.add_discrete([(n, i)`

File: map_coloring.py [hunk 14]
```diff
@@ -69,45 +69,51 @@ def build_graph(state_neighbors):
...
+  
+    # Build the constraints: edges have different color end points
     for u, v in G.edges:
-        dqm.set_quadratic(u, v, {(c, c): lagrange for c in colors})
+        for i in range(num_colors):
+            c = BinaryQuadraticModel('BINARY')
+            c.set_quadratic((u,i),(v,i),1)
```
Comment (JoelPasvolsky) [inline map_coloring.py:95]: `c.set_quadratic((u,i),(v,i),1)` --> `c.set_quadratic((u, i), (v, i), 1)`

File: map_coloring.py [hunk 15]
```diff
@@ -69,45 +69,51 @@ def build_graph(state_neighbors):
...
+            cqm.add_constraint(c == 0)
 
-    return dqm
+    return cqm
 
-def run_hybrid_solver(dqm):
-    """Solve DQM using hybrid solver through the cloud."""
+def run_hybrid_solver(cqm):
+    """Solve CQM using hybrid solver through the cloud."""
```
Comment (JoelPasvolsky) [inline map_coloring.py:101]: :-) 
How about "Solve CQM using a Leap hybrid solver" ?

File: map_coloring.py [hunk 16]
```diff
@@ -69,45 +69,51 @@ def build_graph(state_neighbors):
...
+    # Solve the problem using the CQM solver
+    sampleset = sampler.sample_cqm(cqm, label='Example - Map Coloring')
+    feasible_sampleset = sampleset.filter(lambda row: row.is_feasible)
 
-    # Solve the problem using the DQM solver
-    sampleset = sampler.sample_dqm(dqm, label='Example - Map Coloring')
+    ss = feasible_sampleset.first.sample
 
-    return sampleset
+    soln = {key[0]: key[1] for key, val in ss.items() if val == 1.0}
```
Comment (JoelPasvolsky) [inline map_coloring.py:114]: Same comment as above

File: map_coloring.py [hunk 17]
```diff
@@ -161,12 +167,9 @@ def plot_map(sample, state_records, colors):
     colors = ['purple', 'blue', 'green', 'yellow', 'orange', 'red', 'grey']
```
Comment (JoelPasvolsky) [inline map_coloring.py:167]: If the solution is to use more colors than strictly needed, this should be noted. But why not limit the colors to 4?

**Review by JoelPasvolsky — COMMENTED** [review]
Thanks for calling the PR to my attention, I missed the original request, sorry.

File: README.md [hunk 18]
```diff
@@ -1,127 +1,97 @@
...
+the state of Maryland should be colored with color 0 in the USA map. If
+variable `(n, i)` equals 1 then we assign node `n` color `i`.
+
+### Objective
+
+We set the objective in order to use as few colors as possible. We assume a
+very simple relationship: variable `(node, k)` is given a coefficient of k in
+our objective function. This will encourage the solver to use the lowest
+numbered colors as much as possible by minimizing the overall number of colors
+used.
```
[the remark this answers is not in the document]
Comment (vgoliber) [reply README.md:61]: I wasn't the original author on this so didn't come up with the formulation, but I can change the code to set 4 colors as a requirement.

File: map_coloring.py [hunk 19]
```diff
@@ -69,45 +69,48 @@ def build_graph(state_neighbors):
...
+            c.set_quadratic((u, i), (v, i), 1)
+            cqm.add_constraint(c == 0)
 
-    return dqm
+    return cqm
 
-def run_hybrid_solver(dqm):
-    """Solve DQM using hybrid solver through the cloud."""
+def run_hybrid_solver(cqm):
+    """Solve CQM using LeapHybridCQMSampler through the cloud."""
```
Comment (JoelPasvolsky) [inline map_coloring.py:94]: It was the "through the cloud" that had struck me as off but it's not important.

File: README.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (JoelPasvolsky) [reply README.md]: Fair enough.

**Review by JoelPasvolsky — APPROVED** [review]
Thanks!

File: graph_coloring.py [hunk 20]
```diff
@@ -12,78 +12,107 @@
...
+
+    print("\nSaving results in {}...".format(fname))
+
+# ------- Main program -------
+if __name__ == "__main__":
+
+    num_nodes = 50
+
+    G, pos = build_graph(num_nodes)
+    num_colors = max(d for _, d in G.degree())-1
```
Comment (mcfarljm) [inline graph_coloring.py:109]: Should this be degree + 1?
```suggestion
    num_colors = max(d for _, d in G.degree()) + 1
```
